modules/openthread/platform/diag.c

```diff
@@ -68,3 +73,151 @@ void otPlatDiagAlarmCallback(otInstance *aInstance)
 {
 	ARG_UNUSED(aInstance);
 }
+
+/*
+ * To enable gpio diag commands, create `openthread-gpio-ports` property
+ * in `zephyr,user` node.
+ * Property should contain phandles of gpio port nodes, e.g:
+ * zephyr,user {
+ *		openthread-gpio-ports = <&gpio0 &gpio1>;
+ * };
+ *
+ * To enable reading current gpio pin mode, define
+ * `CONFIG_GPIO_GET_DIRECTION` in prj.conf.
+ *
+ * Note: `<gpio>` in `diag gpio` commands is an integer that combines
+ * port and pin into a single, contiguous number space as follows:
+ * gpio = (port * 32) + pin
+ */
+#if DT_NODE_EXISTS(DT_PATH(zephyr_user)) && \
+	DT_NODE_HAS_PROP(DT_PATH(zephyr_user), openthread_gpio_ports)
+
+#if DT_PROP_LEN(DT_PATH(zephyr_user), openthread_gpio_ports) == 1
+static const struct device *const gpio_port_dev[] = {
+	DT_GPIO_PORT_FROM_PROP_IDX(DT_PATH(zephyr_user), openthread_gpio_ports, 0)};
+#else
+static const struct device *const gpio_port_dev[] = {DT_FOREACH_PROP_ELEM_SEP(
+	DT_PATH(zephyr_user), openthread_gpio_ports, DT_GPIO_PORT_FROM_PROP_IDX, (, ))};
+#endif /*DT_PROP_LEN(DT_PATH(zephyr_user), openthread_gpio_ports) == 1*/
+
+const struct device *const otGpioGetDevAndPin(uint32_t aGpio, gpio_pin_t *pin, otError *error)
```

Comment:
This should be static? Also, for non-OT defined APIs, please use Zephyr coding style (i. e. snake_case).